src/oidcendpoint/session/manager.py

```diff
@@ -0,0 +1,429 @@
+
+
+def public_id(uid, salt="", **kwargs):
+    return hashlib.sha256("{}{}".format(uid, salt).encode("utf-8")).hexdigest()
+
+
+def ephemeral_id(*args, **kwargs):
+    return uuid.uuid4().hex
+
+
+class SessionManager(Database):
+    def __init__(self, handler: TokenHandler, db: Optional[object] = None,
+                 conf: Optional[dict] = None, sub_func: Optional[dict] = None):
+        Database.__init__(self, db)
+        self.token_handler = handler
+        self.salt = rndstr(32)
+        self.conf = conf or {}
+
+        # this allows the subject identifier minters to be defined by someone
+        # else then me.
+        if sub_func is None:
+            self.sub_func = {
+                "public": public_id,
+                "pairwise": pairwise_id,
+                "ephemeral": ephemeral_id
+            }
+        else:
+            self.sub_func = sub_func
+            if "public" not in sub_func:
+                self.sub_func["public"] = public_id
+            if "pairwise" not in sub_func:
+                self.sub_func["pairwise"] = pairwise_id
+            if "ephemeral" not in sub_func:
+                self.sub_func["ephemeral"] = ephemeral_id
+
+    def get_user_info(self, uid: str) -> UserSessionInfo:
+        usi = self.get([uid])
+        if isinstance(usi, UserSessionInfo):
+            return usi
+        else:
+            raise ValueError("Not UserSessionInfo")
+
+    def find_token(self, session_id: str, token_value: str) -> Optional[Token]:
+        """
+
+        :param session_id: Based on 3-tuple, user_id, client_id and grant_id
+        :param token_value:
+        :return:
+        """
+        user_id, client_id, grant_id = unpack_session_key(session_id)
+        grant = self.get([user_id, client_id, grant_id])
+        for token in grant.issued_token:
+            if token.value == token_value:
+                return token
+
+        return None
+
+    def create_grant(self,
+                     authn_event: AuthnEvent,
+                     auth_req: AuthorizationRequest,
+                     user_id: str,
+                     client_id: Optional[str] = "",
+                     sub_type: Optional[str] = "public",
+                     token_usage_rules: Optional[dict] = None) -> str:
+        """
+
+        :param authn_event:
+        :param auth_req:
+        :param user_id:
+        :param client_id:
+        :param sub_type:
+        :param token_usage_rules:
+        :return:
+        """
+        try:
+            sector_identifier = auth_req.get("sector_identifier_uri")
+        except AttributeError:
+            sector_identifier = ""
+
+        grant = Grant(authorization_request=auth_req,
+                      authentication_event=authn_event,
+                      sub=self.sub_func[sub_type](
+                          user_id, salt=self.salt,
+                          sector_identifier=sector_identifier),
+                      usage_rules=token_usage_rules
+                      )
```

Comment:
Add scopes to the grant.

Reply:
So, my thinking was that you first get a grant instance with a basic set of attributes (scopes and claims are not among those) and then you add to that after you've dealt with user consent and what ever other policy processes you have to go through.

Reply:
I guess you could add the option to add scopes to grant in creation (through a scopes param in `create_session` and `create_grant` which would default to False). My thinking is that when using a database each call is a query so I would want to make as few as possible when creating a grant.

Since in some use cases you already know the scopes/claims you want to release when the user logs in, it makes sense to provide this option

Reply:
I can definitely buy the optimisation argument and having a scopes parameter in create_session/create_grant would not be the end of the world.
I agree that in some use cases you know what you want to release but then there is user consent and you may have to change the grant anyway. Granted it may not happen every time so we'll save occasional calls to the database.
OK, I'm sold I'll add the parameters.